File: src/main/java/com/example/the_machine/domain/RunEntity.java
package com.example.the_machine.domain;

import com.fasterxml.jackson.databind.JsonNode;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import java.time.Instant;
import java.util.UUID;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

@Entity
@Table(name = "run")
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
public class RunEntity {

  @Id
  private UUID id;

  @Column(nullable = false)
  private UUID threadId;

  @Column(nullable = false)
  private UUID assistantId;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private Status status;

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  private Kind kind;

  private String reason;

  private Instant startedAt;

  private Instant completedAt;

  @Column(columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  private JsonNode usageJson;

  private String error;

  public enum Status {
    QUEUED, RUNNING, REQUIRES_ACTION, COMPLETED, FAILED, CANCELLED
  }

  public enum Kind {
    FULL, PARTIAL
  }
}

File: src/main/kotlin/com/example/the_machine/domain/RunEntity.kt
package com.example.the_machine.domain

import com.example.the_machine.common.JsonElementConverter
import jakarta.persistence.*
import kotlinx.serialization.json.JsonElement
import org.hibernate.annotations.JdbcTypeCode
import org.hibernate.type.SqlTypes
import java.time.Instant
import java.util.*

@Entity
@Table(name = "run")
data class RunEntity(
  @Column(nullable = false)
  val threadId: UUID,

  @Column(nullable = false)
  val assistantId: UUID,

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  val status: Status,

  @Enumerated(EnumType.STRING)
  @Column(nullable = false)
  val kind: Kind,

  val reason: String? = null,

  val startedAt: Instant? = null,

  val completedAt: Instant? = null,

  @Column(columnDefinition = "jsonb")
  @JdbcTypeCode(SqlTypes.JSON)
  @Convert(converter = JsonElementConverter::class)
  val usageJson: JsonElement? = null,

  val error: String? = null
) : BaseEntity() {

  enum class Status {
    QUEUED, RUNNING, REQUIRES_ACTION, COMPLETED, FAILED, CANCELLED
  }

  enum class Kind {
    FULL, PARTIAL
  }
}